core/src/send_transaction_service.rs

```diff
@@ -0,0 +1,387 @@
+use solana_sdk::{clock::Slot, pubkey::Pubkey, signature::Signature};
+use std::sync::Mutex;
+use std::{
+    collections::HashMap,
+    net::{SocketAddr, UdpSocket},
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        mpsc::Receiver,
+        Arc, RwLock,
+    },
+    thread::{self, Builder, JoinHandle},
+    time::{Duration, Instant},
+};
+
+/// Maximum size of the transaction queue
+const MAX_TRANSACTION_QUEUE_SIZE: usize = 10_000; // This seems like a lot but maybe it needs to be bigger one day
+
+pub struct SendTransactionService {
+    thread: JoinHandle<()>,
+}
+
+pub struct TransactionInfo {
+    pub signature: Signature,
+    pub wire_transaction: Vec<u8>,
+    pub last_valid_slot: Slot,
+}
+
+impl TransactionInfo {
+    pub fn new(signature: Signature, wire_transaction: Vec<u8>, last_valid_slot: Slot) -> Self {
+        Self {
+            signature,
+            wire_transaction,
+            last_valid_slot,
+        }
+    }
+}
+
+#[derive(Default, Debug, PartialEq)]
+struct ProcessTransactionsResult {
+    rooted: u64,
+    expired: u64,
+    retried: u64,
+    failed: u64,
+    retained: u64,
+}
+
+impl SendTransactionService {
+    pub fn new(
+        tpu_address: SocketAddr,
+        bank_forks: &Arc<RwLock<BankForks>>,
+        leader_info: Option<(Arc<ClusterInfo>, Arc<Mutex<PohRecorder>>)>,
+        exit: &Arc<AtomicBool>,
+        receiver: Receiver<TransactionInfo>,
+    ) -> Self {
+        let thread = Self::retry_thread(
+            tpu_address,
+            receiver,
+            bank_forks.clone(),
+            leader_info,
+            exit.clone(),
+        );
+        Self { thread }
+    }
+
+    fn retry_thread(
+        tpu_address: SocketAddr,
+        receiver: Receiver<TransactionInfo>,
+        bank_forks: Arc<RwLock<BankForks>>,
+        leader_info: Option<(Arc<ClusterInfo>, Arc<Mutex<PohRecorder>>)>,
+        exit: Arc<AtomicBool>,
+    ) -> JoinHandle<()> {
+        let mut last_status_check = Instant::now();
+        let mut transactions = HashMap::new();
+        let send_socket = UdpSocket::bind("0.0.0.0:0").unwrap();
+
+        let mut recent_peers: HashMap<Pubkey, SocketAddr> = leader_info
+            .as_ref()
+            .map(|cluster_info| {
+                cluster_info
+                    .0
+                    .tpu_peers()
+                    .into_iter()
+                    .map(|ci| (ci.id, ci.tpu))
+                    .collect()
+            })
+            .unwrap_or_default();
+
+        Builder::new()
+            .name("send-tx-svc".to_string())
+            .spawn(move || loop {
+                if exit.load(Ordering::Relaxed) {
+                    break;
+                }
+
+                if let Ok(transaction_info) = receiver.recv_timeout(Duration::from_secs(1)) {
+                    let address = leader_info
+                        .as_ref()
+                        .and_then(|recorder| recorder.1.lock().unwrap().leader_after_n_slots(0))
```

Comment:
Should we aim at the next slot instead?
```suggestion
                        .and_then(|recorder| recorder.1.lock().unwrap().leader_after_n_slots(1))
```

Reply:
Why?  Current leader should be the right slot unless they are done and they will send it to the next leader.

Reply:
Half the time you'll be in the last 50% of the current slot when this code runs, then include the network propagation time from RPC to the leader -- this takes us to very close to the end of the slot or into the next slot already by the time it reaches the target node.

If we miss the leader, yes they will send it to the next leader.
But if we hit the leader one slot early, they'll just hang on to the transaction.

EmpIrically I think this means that we're more likely to hit the actual leader and avoid the forward.

Reply:
I think we should fanout the transaction to the current + next few (2-3) leaders

Reply:
Ok.  I don't think we need to hold up this PR due to this thread.  The current patch is a definite improvement over the current situation, we can iterate more later if needed